ARROW-16782: [Format] Add REE definitions to FlatBuffers (#14176)
apache/arrow: docs/source/format/Columnar.rst

We discuss dictionary encoding as it relates to serialization further
below.

.. _run-end-encoded-layout:

Run-End Encoded Layout
-------------------------

> Run-End is a data representation that represents data as sequences of the
> same value, called runs. Each run is represented as a value, and an integer
> describing the index in the array where the run ends.

Reviewer: ```suggestion
Run-end encoding (REE) is a variation of run-length encoding (RLE). Run-end
encoding is well-suited for representing data containing sequences of the same
value, called runs. In run-end encoding, each run is represented as a value
and an integer giving the index in the array where the run ends.
```

Reviewer: It looks like this and some other changes were reverted in the rebase

Reviewer 2: yea, i noticed. I'm fixing it.

Any array can be run-end encoded. A run-end encoded array has no buffers
by itself, but has two child arrays. The first one holds a signed integer
called a "run end" for each run. The run ends array can hold either 16, 32, or
64-bit integers. The actual values of each run are held
the second child array.

The values in the first child array represent the length of each run. They do
not hold the length of the respective run directly, but the accumulated length
of all runs from the first to the current one, i.e. the logical index where the
current run ends. This allows relatively efficient random access from a logical

> index using binary search. The length of an individual run can be determined by
> subtracting two adjacent values.

Reviewer: ```suggestion
index using binary search. The length of an individual run can be determined by
subtracting two adjacent values. (Contrast this with run-length encoding, in
which the lengths of the runs are represented directly, and in which random
access is less efficient.)
```

We discuss dictionary encoding as it relates to serialization further
below.

.. _run-end-encoded-layout:

Run-End Encoded Layout
-------------------------

Run-end encoding (REE) is a variation of run-length encoding (RLE). These
encodings are well-suited for representing data containing sequences of the
same value, called runs. In run-end encoding, each run is represented as a
value and an integer giving the index in the array where the run ends.

Any array can be run-end encoded. A run-end encoded array has no buffers
by itself, but has two child arrays. The first one holds a signed integer
called a "run end" for each run. The run ends array can hold either 16, 32, or

> 64-bit integers. The actual values of each run are held
> the second child array.

Reviewer 3: ```suggestion
64-bit integers. The actual values of each run are held in
the second child array.
```

We discuss dictionary encoding as it relates to serialization further
below.

.. _run-end-encoded-layout:

> Run-End Encoded Layout

Reviewer 2: Interestingly dictionary encoding isn't listed in that list of physical layouts, and REE is probably closest to that in terms of definitions. Should we also add dictionary encoded layout to this list?

> Run-End Encoded Layout

Reviewer 4: Dictionary encoding is a bit different in the specification since it's treated separately and not as a "Type", that's maybe the reason it's not listed there.

> Run-End Encoded Layout

Reviewer 2: I added both to that list, let me know what you think of the addition and if you think I should tweak it at all.

> Run-End Encoded Layout
> -------------------------

Reviewer 5: You need to ensure that the decoration is exactly aligned to the title:
```suggestion
Run-End Encoded Layout
----------------------
```

Run-end encoding (REE) is a variation of run-length encoding (RLE). These
encodings are well-suited for representing data containing sequences of the
same value, called runs. In run-end encoding, each run is represented as a
value and an integer giving the index in the array where the run ends.

Any array can be run-end encoded. A run-end encoded array has no buffers

> by itself, but has two child arrays. The first one holds a signed integer
> called a "run end" for each run. The run ends array can hold either 16, 32, or
> 64-bit integers. The actual values of each run are held in the second child array.

Reviewer 5: There seems to be some redundancy here
```suggestion
by itself, but has two child arrays. The first child array, called the run ends array,
holds either 16, 32, or 64-bit signed integers. The actual values of each run
are held in the second child array.
```

For the purposes of determining field names and schemas, these child arrays
are prescribed the standard names of **run_ends** and **values** respectively.

The values in the first child array represent the length of each run. They do
not hold the length of the respective run directly, but the accumulated length
of all runs from the first to the current one, i.e. the logical index where the

> current run ends. This allows relatively efficient random access from a logical

Reviewer 5: This is confusing. Why start by saying they represent the length, and then explain that they don't hold the actual length?

Reviewer 2: Is the updated version better?

Reviewer 5: Yes, thanks!

index using binary search. The length of an individual run can be determined by
subtracting two adjacent values. (Contrast this with run-length encoding, in
which the lengths of the runs are represented directly, and in which random
access is less efficient.)

A run must have have a length of at least 1. This means the values in the

> run ends array all positive and in strictly ascending order. A run end cannot be

Reviewer 5: ```suggestion
run ends array all are positive and in strictly ascending order. A run end cannot be
```

null.

As an example, you could have the following data: ::

    type: Float32
    [1.0, 1.0, 1.0, 1.0, null, null, 2.0]

In Run-end-encoded form, this could appear as:

::

    * Length: 7, Null count: 2

>     * Children arrays:

Reviewer 5: Is it "children arrays" or "child arrays"?

Reviewer 2: updated to "Child Arrays"

      * run_ends (Int32):
        * Length: 3, Null count: 0

>         * Validity bitmap buffer: Not required

Reviewer 4: The run ends can have nulls?

Reviewer 2: there shouldn't be any nulls in the run_ends array, i'll make that explicit. Good catch

Reviewer 2: Actually it's already stated just above these examples that a run end cannot be null, i'll just update this diagram to be explicit.

Reviewer 4: That might have been discussed before(and probably also part of the vote), but if they can't have nulls and thus basically are a single buffer, is there a reason that run_ends are not a "buffer" of the array instead of the first "child array" ? (like variable size list has offsets as buffer and values as child array)

Reviewer 4: (have them as a buffer would avoid the ambiguity around the validity bitmap of the run_ends as child array)

Reviewer 2: found the mailing list discussion on this which explicitly addresses being children arrays rather than a buffer: https://lists.apache.org/thread/bfz3m5nyf7flq7n6q9b1bx3jhcn4wq29 

TL;DR: in addition to the point I made above, there's this comment from @westonpace at the end of the thread:

> * There are two child arrays not necessarily for slicing but more to help
distinguish between the logical length (there are no buffers with the
logical length but there is an array) and the physical length (all of the
buffers are in these child arrays which are sized to the physical length).

Reviewer 3: It might be good to preserve that reasoning somewhere more easily findable for posterity?

Reviewer 2: added to `Columnar.rst`. Let me know if that's sufficient or if anyone thinks I should modify it a bit.

        * Values buffer

>           | Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | Bytes  6-63           |

Reviewer 5: ```suggestion
          | Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | Bytes 12-63           |
```

          |-------------|-------------|-------------|-----------------------|
          | 4           | 6           | 7           | unspecified (padding) |

>     * values (Float32):

Reviewer 5: This should indented the same as "run_ends" above.

Reviewer 2: fixed

        * Length: 3, Null count: 1
        * Validity bitmap buffer:

          | Byte 0 (validity bitmap) | Bytes 1-63            |
          |--------------------------|-----------------------|
          | 00000101                 | 0 (padding)           |

        * Values buffer

>           | Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | Bytes  6-63           |

Reviewer 5: ```suggestion
          | Bytes 0-3   | Bytes 4-7   | Bytes 8-11  | Bytes 12-63           |
```

encodings are well-suited for representing data containing sequences of the
same value, called runs. In run-end encoding, each run is represented as a
value and an integer giving the index in the array where the run ends.

Any array can be run-end encoded. A run-end encoded array has no buffers
by itself, but has two child arrays. The first child array, called the run ends array,
holds either 16, 32, or 64-bit signed integers. The actual values of each run
are held in the second child array.
For the purposes of determining field names and schemas, these child arrays
are prescribed the standard names of **run_ends** and **values** respectively.

The values in the first child array represent the accumulated length of all runs 
from the first to the current one, i.e. the logical index where the
current run ends. This allows relatively efficient random access from a logical
index using binary search. The length of an individual run can be determined by
subtracting two adjacent values. (Contrast this with run-length encoding, in
which the lengths of the runs are represented directly, and in which random
access is less efficient.) 

.. note::

>    Because the ``run_ends`` child array cannot have nulls, it's reasonable
>    to consider why the ``run_ends`` are a child array instead of just a
>    buffer, like the offsets for a :ref:`variable-size-list-layout`. This
>    layout was considered, but it was decided to use the child arrays. 
> 
>    The full discussion can be found in the `mailing list REE layout discussion`_
>    but the gist of it is the following:
> 
>    While maintaining two child arrays makes it easier and more efficient to
>    handle slicing run-end encoded arrays, the decision was more to help
>    distinguish between the logical length (there are no buffers with the
>    logical length, but there is an array) and the physical length (all of
>    the buffers are in these child arrays which are sized to the physical length).

Reviewer 6: I don't know if we want to justify our decision quite so much on the spec page.  Maybe something simpler like:

```suggestion
   Because the ``run_ends`` child array cannot have nulls, it's reasonable
   to consider why the ``run_ends`` are a child array instead of just a
   buffer, like the offsets for a :ref:`variable-size-list-layout`. This
   layout was considered, but it was decided to use the child arrays. 

   Child arrays allow us to keep the "logical length" (the decoded length)
   associated with the parent array and the "physical length" (the number
   of run ends) associated with the child arrays.  If ``run_ends`` was a
   buffer in the parent array then the size of the buffer would be unrelated
   to the length of the array and this would be confusing.
```

Reviewer 5: +1, perhaps we need some kind of separate FAQ

Reviewer 2: you think it's not worth keeping the link to the original discussion in the spec?

Reviewer 5: We generally don't keep links to design discussions in the specs themselves, it would be weird to make a special case here.

Reviewer 2: sounds good to me. Committing this suggestion.
